lib/Sema/ConstraintSystem.cpp

```diff
@@ -868,16 +867,36 @@ Type ConstraintSystem::openOpaqueType(Type type,
   return underlyingTyVar;
 }
 
-Type ConstraintSystem::openOpaqueTypeRec(Type type,
-                                         ConstraintLocatorBuilder locator) {
+Type ConstraintSystem::openOpaqueType(Type type, ContextualTypePurpose context,
```

Comment:
I'm not sure why you need to check anything about the `context`. I deleted all `context` checks locally and only kept the `dyn_cast<AbstractFunctionDecl>` check and this still passed the new test case.

If there is a case where checking `context` is necessary--and I certainly could be missing something here--, we need another test that covers that.

Reply:
Because although declaration context might be a function it doesn't mean that it's connected to its result type, that's how the original check was structured as well.

Reply:
I see. I still think it would be a good idea to devise a test that fails without this check.

Reply:
@willtunnels Would be great if you did that in a separate PR, you know more then I do about corner cases for nested opaque types.